pages/change-password/_token.vue

```diff
@@ -0,0 +1,77 @@
+<template>
+  <div>
+    <Portal to="header">
+      <Logo class="mx-auto h-20 w-auto" />
+      <h2 class="mt-8 text-center text-5xl font-extrabold text-gray-900">
+        Change Password
+      </h2>
+    </Portal>
+    <form @submit.prevent="changePassword">
+      <div class="space-y-5">
+        <t-input-group label="New Password" variant="important">
+          <PasswordInput v-model="password" />
+        </t-input-group>
+        <t-input-group label="Confirm Password" variant="important">
+          <PasswordInput v-model="repeatPassword" />
+        </t-input-group>
+      </div>
+      <div class="py-2">
+        <t-button class="w-full" type="submit">Change Password</t-button>
+      </div>
+    </form>
+  </div>
+</template>
+
+<script>
+import {
+  defineComponent,
+  useRouter,
+  useRoute,
+  computed,
+} from '@nuxtjs/composition-api'
+import { ref } from '@vue/composition-api'
+import { gql } from 'graphql-tag'
+import { useChangePasswordMutation } from '../../apollo/graphql'
+export default defineComponent({
+  name: 'ChangePassword',
+  layout: 'bare',
+  setup() {
+    const password = ref('')
+    const repeatPassword = ref('')
+    gql`
+      mutation ChangePassword(
+        $token: String!
+        $email: String!
+        $newPassword: String!
+      ) {
+        changePassword(
+          token: $token
+          email: $email
+          newPassword: $newPassword
+        ) {
+          id
+        }
+      }
+    `
+    const route = useRoute()
+    const token = computed(() => route.value.query.token)
+    const email = computed(() => route.value.query.email)
```

Comment:
Earlier implementation has some downsides 
1. Users can have multiple windows open to reset the password.
2. Hashing token is not possible because it is being used as a key.

To address the first issue, I'd suggest that we ask the users to put in the email with a new password, which also fixes the second issue. 

For the second issue:
We can use a secret prefix(random string) for the token, store it as a key, use email as value, and store another key-value pair with key as email and value as isWindowOpen.

Reply:
These are good points. What about using the userId instead of the email? Then we should get both approaches, but someone stealing the link still cannot login unless he knows the corresponding email? I guess it should be even possible to use a hashed version of the userid, so that even if one gets the "userid" from the query, one still cannot guess the identity of the user.

I went through some of the reset-password emails I got in the past, and most of them use either a simple token (discourse) or userid + token (twitter, facebook [not sure, they send `u` and `n` params], pinterest, skype).

Reply:
Userid is a good option. But the issue with userid is that it can't be hashed if we use it as a key. 
I think there shouldn't be an issue in exposing userid.

Reply:
I guess you could add an additional hashed userid <-> userid mapping in redis, but I agree that's not worth the overhead.